crowbar_framework/app/models/service_object.rb

```diff
@@ -649,6 +645,33 @@ def find_dep_proposal(bc, optional=false)
     proposals[0] || ""
   end
 
+  def node_is_valid_for_role(node, role)
+    elements = { role => [node] }
+    violates_admin_constraint?(elements, role) ||
+      violates_platform_constraint?(elements, role) ||
+      violates_exclude_platform_constraint?(elements, role) ||
+      violates_cluster_constraint?(elements, role)
+  end
+
+  # Helper to select nodes that make sense on proposal creation
+  def select_nodes_for_role(all_nodes, role, preferred_intended_role = nil)
+    # do not modify array given by caller
+    all_nodes = all_nodes.dup
```

Comment:
Just a nitpicking - what about different name than all_nodes, when it's actually be a subset anyway...